### PR TITLE
Link with price

### PR DESCRIPTION
**Pull request by anxolin**
* Adapts the bot "fill order URL" generation to use the price instead of the receiving amount
* It handles unlimited orders: Calculates the prices making the taker absorb the fee so the two trades can match
* It handles limited orders: Calculates the amount required to full fill the maker's order (taking the fee into account), and then calculates the price depending on that
* Both price calculations takes the token precision into account, otherwise it's possible the orders don't match because of rounding issues

Example:
image

Additionally:
* The partials that compose the final message were broken down into smaller units
* Added comments and examples to make the message generation more readble
* Added some test cases: Edge cases related to rounding issues
* Add GIVEN/WHEN/THEN pattern to some test for readability
* Mock time in tests to be able to test using equals for example in relative times. i.e. `expect(active).toEqual('`Today at 3:10 AM GMT`, `in 3 hours`')` this way the test is more readable and is still repeatable.

### CODE REVIEW

File: src/helpers/message.ts [hunk 1]
```diff
@@ -27,50 +28,56 @@ function _getTokenFmt(amount: BigNumber, token: TokenDto) {
...
+): BigNumber {
   const { buyToken, sellToken, priceNumerator, priceDenominator } = order
+  const buyTokenDecimals = buyToken.decimals
+  const sellTokenDecimals = sellToken.decimals
 
-  if (buyToken.decimals >= sellToken.decimals) {
-    const precisionFactor = 10 ** (buyToken.decimals - sellToken.decimals)
-    return priceNumerator.dividedBy(priceDenominator.multipliedBy(precisionFactor))
+  let price
+  if (buyTokenDecimals >= sellToken.decimals) {
```
Comment (alfetopito) [inline src/helpers/message.ts]:
```suggestion
  if (buyTokenDecimals >= sellTokenDecimals) {
```

File: src/helpers/message.ts [hunk 2]
```diff
@@ -27,50 +28,56 @@ function _getTokenFmt(amount: BigNumber, token: TokenDto) {
...
+    const precisionFactor = 10 ** (buyTokenDecimals - sellTokenDecimals)
+    price = priceNumerator.dividedBy(priceDenominator.multipliedBy(precisionFactor))
   } else {
-    const precisionFactor = 10 ** (sellToken.decimals - buyToken.decimals)
-    return priceNumerator.multipliedBy(precisionFactor).dividedBy(priceDenominator)
+    const precisionFactor = 10 ** (sellTokenDecimals - buyTokenDecimals)
+    price = priceNumerator.multipliedBy(precisionFactor).dividedBy(priceDenominator)
   }
+
+  return price.decimalPlaces(buyTokenDecimals, BigNumber.ROUND_FLOOR)
```
Comment (alfetopito) [inline src/helpers/message.ts]: Is it not a problem to adjust the price only based on `buyTokenDecimals`?

File: src/helpers/message.ts [hunk 3]
```diff
@@ -157,25 +191,55 @@ export function newOrderMessage(order: OrderDto, baseUrl: string): string {
...
 
   // TODO: Should we publish even if the user doesn't have balance. Should we include the balance of the user? he can change it...
   //  https://github.com/gnosis/dex-telegram/issues/45
 
-  const sellMsg = buildSellMsg(isUnlimited, buyTokenLabel, sellTokenLabel, buyAmountFmt, sellAmountFmt)
-  // Calculate the price
+  // Partial message: Order description
+  //    i.e Sell 3.535 WETH for 1,000 SNX
+  const isUnlimited = isOrderUnlimited(priceNumerator, priceDenominator)
+  const sellMsg = buildSellMsg(isUnlimited, buyTokenLabel, sellTokenLabel, buyAmountFmt, sellAmountFmt) + '\n'
```
Comment (Velenir) [inline src/helpers/message.ts]: Would be nice to have object as params for `buildSellMsg`

File: src/helpers/message.ts (hunk 2 above)
Comment (anxolin) [reply src/helpers/message.ts]: At the beginning I was using both, taking the maximum of buyTokenDecimals, sellTokenDecimals.

Then I realised that the price is used to calculate the receiving amount, that are expressed in buy tokens. So I've used the precision of the buy token and round down (slightly worse price)

Comment (anxolin) [reply src/helpers/message.ts]: Now that I think about this again, I'm not 100% sure of the decimals we can have in the price.

I'm starting to think that my approach could fail with sell amounts with decimals.

Let me formulate it:
* We are selling `X` tokens for `Y` tokens at a price `P` => `P x X = Y`
* `precision(x)` is the precision of the token `x`
* `decimals(x)` is the number of decimals in `x`

We know:
* `decimals(x) < precision(x), for X and Y`

Ideally we want to find a P:
* That `decimals(P x X) <= precision(Y)`

Comment (anxolin) [reply src/helpers/message.ts]: Worst case is:
`decimals(P x X) = decimals(P) +  decimals(X)`

Therefore, there's not such `P` 😔

...unless we do this price calculation in two steps:

Let's see an example:
```
* decimals(A) = 2
* decimals(A) = 1
* amount A = 1.23

P = 9.87654321
P x A = 1.23 * 9.837398374 = 12.1481481483
B = 12.1

Then we adjust P to satisfy:
1.23 * P = 12.1
P = 12.1 / 1.23 = 9.837398374
```

Comment (anxolin) [reply src/helpers/message.ts]: I'll make a separate PR for this

File: src/helpers/message.ts (hunk 3 above)
Comment (anxolin) [reply src/helpers/message.ts]: 40457d71e83aa649b8ebdf7e90b231ac692e1a49
